README.md
# Bypass

<!-- MDOC !-->

[![Build Status](https://travis-ci.org/PSPDFKit-labs/bypass.svg?branch=master)](https://travis-ci.org/PSPDFKit-labs/bypass)
[![Module Version](https://img.shields.io/hexpm/v/bypass.svg)](https://hex.pm/packages/bypass)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/bypass/)
[![Total Download](https://img.shields.io/hexpm/dt/bypass.svg)](https://hex.pm/packages/bypass)
[![License](https://img.shields.io/hexpm/l/bypass.svg)](https://github.com/PSPDFKit-labs/bypass/blob/master/LICENSE)
[![Last Updated](https://img.shields.io/github/last-commit/PSPDFKit-labs/bypass.svg)](https://github.com/PSPDFKit-labs/bypass/commits/master)


`Bypass` provides a quick way to create a custom plug that can be put in place
instead of an actual HTTP server to return prebaked responses to client
requests. This is most useful in tests, when you want to create a mock HTTP
server and test how your HTTP client handles different types of responses from
the server.

Bypass supports Elixir 1.7 and OTP 20 and up. It works with Cowboy 2.

## Usage

To use Bypass in a test case, open a connection and use its port to connect your
client to it.

If you want to test what happens when the HTTP server goes down, use
`Bypass.down/1` to close the TCP socket and `Bypass.up/1` to start listening on
the same port again. Both functions block until the socket updates its state.

### Expect Functions

You can take any of the following approaches:

* `expect/2` or `expect_once/2` to install a generic function that all calls to
  bypass will use
* `expect/4` and/or `expect_once/4` to install specific routes (method and path)
* `stub/4` to install specific routes without expectations
* a combination of the above, where the routes will be used first, and then the
  generic version will be used as default

### Example

In the following example `TwitterClient.start_link()` takes the endpoint URL as
its argument allowing us to make sure it will connect to the running instance of
Bypass.

```elixir
defmodule TwitterClientTest do
  use ExUnit.Case, async: true

  setup do
    bypass = Bypass.open()
    {:ok, bypass: bypass}
  end

  test "client can handle an error response", %{bypass: bypass} do
    Bypass.expect_once(bypass, "POST", "/1.1/statuses/update.json", fn conn ->
      Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
    end)

    {:ok, client} = TwitterClient.start_link(url: endpoint_url(bypass.port))
    assert {:error, :rate_limited} == TwitterClient.post_tweet(client, "Elixir is awesome!")
  end

  test "client can recover from server downtime", %{bypass: bypass} do
    Bypass.expect(bypass, fn conn ->
      # We don't care about `request_path` or `method` for this test.
      Plug.Conn.resp(conn, 200, "")
    end)

    {:ok, client} = TwitterClient.start_link(url: endpoint_url(bypass.port))

    assert :ok == TwitterClient.post_tweet(client, "Elixir is awesome!")

    # Blocks until the TCP socket is closed.
    Bypass.down(bypass)

    assert {:error, :noconnect} == TwitterClient.post_tweet(client, "Elixir is awesome!")

    Bypass.up(bypass)

    # When testing a real client that is using e.g. https://github.com/fishcakez/connection
    # with https://github.com/ferd/backoff to handle reconnecting, we'd have to loop for
    # a while until the client has reconnected.

    assert :ok == TwitterClient.post_tweet(client, "Elixir is awesome!")
  end

  defp endpoint_url(port), do: "http://localhost:#{port}/"
end
```

That's all you need to do. Bypass automatically sets up an `on_exit` hook to
close its socket when the test finishes running.

Multiple concurrent Bypass instances are supported, all will have a different
unique port.  Concurrent requests are also supported on the same instance.

> Note: `Bypass.open/0` **must not** be called in a `setup_all` blocks due to
> the way Bypass verifies the expectations at the end of each test.

## How to use with ESpec

While Bypass primarily targets ExUnit, the official Elixir builtin test
framework, it can also be used with [ESpec](https://hex.pm/packages/espec). The
test configuration is basically the same, there are only two differences:

1. In your Mix config file, you must declare which test framework Bypass is
   being used with (defaults to `:ex_unit`). This simply disables the automatic
   integration with some hooks provided by `ExUnit`.

   ```elixir
   config :bypass, test_framework: :espec
   ```

2. In your specs, you must explicitly verify the declared expectations. You can
   do it in the `finally` block.

   ```elixir
   defmodule TwitterClientSpec do
     use ESpec, async: true

     before do
       bypass = Bypass.open()
       {:shared, bypass: bypass}
     end

     finally do
       Bypass.verify_expectations!(shared.bypass)
     end

     specify "the client can handle an error response" do
       Bypass.expect_once(shared.bypass, "POST", "/1.1/statuses/update.json", fn conn ->
         Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
       end)

       {:ok, client} = TwitterClient.start_link(url: endpoint_url(shared.bypass.port))
       assert {:error, :rate_limited} == TwitterClient.post_tweet(client, "Elixir is awesome!")
     end

     defp endpoint_url(port), do: "http://localhost:#{port}/"
   end
   ```

## Configuration options

Set `:enable_debug_log` to `true` in the application environment to make Bypass
log what it's doing:

```elixir
config :bypass, enable_debug_log: true
```

<!-- MDOC !-->

## Installation

Add `:bypass` to your list of dependencies in mix.exs:

```elixir
def deps do
  [
    {:bypass, "~> 2.1", only: :test}
  ]
end
```

We do not recommended adding `:bypass` to the list of applications in your
`mix.exs`.

## License

This software is licensed under [the MIT license](LICENSE).

## About

<a href="https://pspdfkit.com/">
  <img src="https://avatars2.githubusercontent.com/u/1527679?v=3&s=200" height="80" />
</a>

This project is maintained and funded by [PSPDFKit](https://pspdfkit.com/).

Please ensure [you signed our
CLA](https://pspdfkit.com/guides/web/current/miscellaneous/contributing/) so we
can accept your contributions.

See [our other open source projects](https://github.com/PSPDFKit-labs), read
[our blog](https://pspdfkit.com/blog/) or say hello on Twitter
([@PSPDFKit](https://twitter.com/pspdfkit)).


CHANGELOG.md
# Changelog

## v2.1.0 - 13 Nov 2020

* Support latest Cowboy.
* Require at least Elixir 1.7.
* Ditch Cowboy 1.0

## v2.0.0 - 19 Aug 2020

* Allow the redefinition of routes.
* Make listen interface configurable.
* Add SO_REUSEPORT.
* Add support for parametric routes.
* Switch from :simple_one_for_one to DynamicSupervisor.
* Require at least Elixir 1.6.
* Replace gun with mint.

## v1.0.0 - 26 Nov 2018

* Support for Plug 1.7 with `plug_cowboy` 1 and 2.

## v0.9.0 - 30 Sept 2018

* Add support for Cowboy 2 thanks to @hassox

## v0.6.0 - 2 Feb 2017

* Add support for Elixir v1.0
* Allow choosing the port number for a Bypass instance
* Bypass instances now only listen on the loopback interface


lib/bypass.ex
defmodule Bypass do
  @external_resource "README.md"
  @moduledoc "README.md"
             |> File.read!()
             |> String.split("<!-- MDOC !-->")
             |> Enum.fetch!(1)

  defstruct pid: nil, port: nil

  @typedoc """
  Represents a Bypass server process.
  """
  @type t :: %__MODULE__{pid: pid, port: non_neg_integer}

  import Bypass.Utils
  require Logger

  @doc """
  Starts an Elixir process running a minimal Plug app. The process is a HTTP
  handler and listens to requests on a TCP port on localhost.

  Use the other functions in this module to declare which requests are handled
  and set expectations on the calls.

  ## Options

  - `port` - Optional TCP port to listen to requests.

  ## Examples

  ```elixir
  bypass = Bypass.open()
  ```

  Assign a specific port to a Bypass instance to listen on:

  ```elixir
  bypass = Bypass.open(port: 1234)
  ```

  """
  @spec open(Keyword.t()) :: Bypass.t()
  def open(opts \\ []) do
    pid = start_instance(opts)
    port = Bypass.Instance.call(pid, :port)
    debug_log("Did open connection #{inspect(pid)} on port #{inspect(port)}")
    bypass = %Bypass{pid: pid, port: port}
    setup_framework_integration(test_framework(), bypass)
    bypass
  end

  defp start_instance(opts) do
    case DynamicSupervisor.start_child(Bypass.Supervisor, Bypass.Instance.child_spec(opts)) do
      {:ok, pid} ->
        pid

      {:ok, pid, _info} ->
        pid

      {:error, reason} ->
        raise "Failed to start bypass instance.\n" <>
                "Reason: #{start_supervised_error(reason)}"
    end
  end

  defp start_supervised_error({{:EXIT, reason}, info}) when is_tuple(info),
    do: Exception.format_exit(reason)

  defp start_supervised_error({reason, info}) when is_tuple(info),
    do: Exception.format_exit(reason)

  defp start_supervised_error(reason), do: Exception.format_exit({:start_spec, reason})

  defp setup_framework_integration(:ex_unit, bypass = %{pid: pid}) do
    ExUnit.Callbacks.on_exit({Bypass, pid}, fn ->
      do_verify_expectations(bypass.pid, ExUnit.AssertionError)
    end)
  end

  defp setup_framework_integration(:espec, _bypass) do
  end

  @doc """
  Can be called to immediately verify if the declared request expectations have
  been met.

  Returns `:ok` on success and raises an error on failure.
  """
  @spec verify_expectations!(Bypass.t()) :: :ok | no_return()
  def verify_expectations!(bypass) do
    verify_expectations!(test_framework(), bypass)
  end

  defp verify_expectations!(:ex_unit, _bypass) do
    raise "Not available in ExUnit, as it's configured automatically."
  end

  if Code.ensure_loaded?(ESpec) do
    defp verify_expectations!(:espec, bypass) do
      do_verify_expectations(bypass.pid, ESpec.AssertionError)
    end
  end

  defp do_verify_expectations(bypass_pid, error_module) do
    case Bypass.Instance.call(bypass_pid, :on_exit) do
      :ok ->
        :ok

      :ok_call ->
        :ok

      {:error, :too_many_requests, {:any, :any}} ->
        raise error_module, "Expected only one HTTP request for Bypass"

      {:error, :too_many_requests, {method, path}} ->
        raise error_module, "Expected only one HTTP request for Bypass at #{method} #{path}"

      {:error, :unexpected_request, {:any, :any}} ->
        raise error_module, "Bypass got an HTTP request but wasn't expecting one"

      {:error, :unexpected_request, {method, path}} ->
        raise error_module,
              "Bypass got an HTTP request but wasn't expecting one at #{method} #{path}"

      {:error, :not_called, {:any, :any}} ->
        raise error_module, "No HTTP request arrived at Bypass"

      {:error, :not_called, {method, path}} ->
        raise error_module,
              "No HTTP request arrived at Bypass at #{method} #{path}"

      {:exit, {class, reason, stacktrace}} ->
        :erlang.raise(class, reason, stacktrace)
    end
  end

  @doc """
  Re-opens the TCP socket on the same port. Blocks until the operation is
  complete.

  ```elixir
  Bypass.up(bypass)
  ```
  """
  @spec up(Bypass.t()) :: :ok | {:error, :already_up}
  def up(%Bypass{pid: pid}),
    do: Bypass.Instance.call(pid, :up)

  @doc """
  Closes the TCP socket. Blocks until the operation is complete.

  ```elixir
  Bypass.down(bypass)
  ```
  """
  @spec down(Bypass.t()) :: :ok | {:error, :already_down}
  def down(%Bypass{pid: pid}),
    do: Bypass.Instance.call(pid, :down)

  @doc """
  Expects the passed function to be called at least once regardless of the route.

  ```elixir
  Bypass.expect(bypass, fn conn ->
    assert "/1.1/statuses/update.json" == conn.request_path
    assert "POST" == conn.method
    Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
  end)
  ```
  """
  @spec expect(Bypass.t(), (Plug.Conn.t() -> Plug.Conn.t())) :: :ok
  def expect(%Bypass{pid: pid}, fun),
    do: Bypass.Instance.call(pid, {:expect, fun})

  @doc """
  Expects the passed function to be called at least once for the specified route (method and path).

  - `method` is one of `["GET", "POST", "HEAD", "PUT", "PATCH", "DELETE", "OPTIONS", "CONNECT"]`

  - `path` is the endpoint.

  ```elixir
  Bypass.expect(bypass, "POST", "/1.1/statuses/update.json", fn conn ->
    Agent.get_and_update(AgentModule, fn step_no -> {step_no, step_no + 1} end)
    Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
  end)
  ```
  """
  @spec expect(Bypass.t(), String.t(), String.t(), (Plug.Conn.t() -> Plug.Conn.t())) :: :ok
  def expect(%Bypass{pid: pid}, method, path, fun),
    do: Bypass.Instance.call(pid, {:expect, method, path, fun})

  @doc """
  Expects the passed function to be called exactly once regardless of the route.

  ```elixir
  Bypass.expect_once(bypass, fn conn ->
    assert "/1.1/statuses/update.json" == conn.request_path
    assert "POST" == conn.method
    Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
  end)
  ```
  """
  @spec expect_once(Bypass.t(), (Plug.Conn.t() -> Plug.Conn.t())) :: :ok
  def expect_once(%Bypass{pid: pid}, fun),
    do: Bypass.Instance.call(pid, {:expect_once, fun})

  @doc """
  Expects the passed function to be called exactly once for the specified route (method and path).

  - `method` is one of `["GET", "POST", "HEAD", "PUT", "PATCH", "DELETE", "OPTIONS", "CONNECT"]`

  - `path` is the endpoint.

  ```elixir
  Bypass.expect_once(bypass, "POST", "/1.1/statuses/update.json", fn conn ->
    Agent.get_and_update(AgentModule, fn step_no -> {step_no, step_no + 1} end)
    Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
  end)
  ```
  """
  @spec expect_once(Bypass.t(), String.t(), String.t(), (Plug.Conn.t() -> Plug.Conn.t())) :: :ok
  def expect_once(%Bypass{pid: pid}, method, path, fun),
    do: Bypass.Instance.call(pid, {:expect_once, method, path, fun})

  @doc """
  Allows the function to be invoked zero or many times for the specified route (method and path).

  - `method` is one of `["GET", "POST", "HEAD", "PUT", "PATCH", "DELETE", "OPTIONS", "CONNECT"]`

  - `path` is the endpoint.

  ```elixir
  Bypass.stub(bypass, "POST", "/1.1/statuses/update.json", fn conn ->
    Agent.get_and_update(AgentModule, fn step_no -> {step_no, step_no + 1} end)
    Plug.Conn.resp(conn, 429, ~s<{"errors": [{"code": 88, "message": "Rate limit exceeded"}]}>)
  end)
  ```
  """
  @spec stub(Bypass.t(), String.t(), String.t(), (Plug.Conn.t() -> Plug.Conn.t())) :: :ok
  def stub(%Bypass{pid: pid}, method, path, fun),
    do: Bypass.Instance.call(pid, {:stub, method, path, fun})

  @doc """
  Makes an expectation to pass.

  ```
  Bypass.expect(bypass, fn _conn ->
    Bypass.pass(bypass)

    assert false
  end)
  """
  @spec pass(Bypass.t()) :: :ok
  def pass(%Bypass{pid: pid}),
    do: Bypass.Instance.call(pid, :pass)

  defp test_framework do
    Application.get_env(:bypass, :test_framework, :ex_unit)
  end
end
